File: leetcode_solutions/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = float("-Inf")
        buy_day = float("Inf")
        sell_day = 0
        
        for price in prices:
            buy_day = min(buy_day, price)
            sell_day = max(buy_day, price)
            
            max_profit = max(max_profit, sell_day - buy_day)
        return max_profit
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {      
        int maxProfit = 0;
        int left = 0;
        int right = 1;
        while(right < prices.length){     
            if(prices[left] < prices[right]){
                int profit = Math.abs(prices[right] -  prices[left]);
                maxProfit = Math.max(maxProfit, profit);
            } else{
                // low price
                left = right;
            }
            
            right ++;
            
        }
        
        
        
        return maxProfit;
        
    }
}